### ISSUE TITLE
cmd/go: incorrect sum for replaced module in either 'mod tidy' or 'mod vendor'

### ISSUE THREAD

**Issue by kevinburke** [golang/go#27868]
Please answer these questions before submitting your issue. Thanks!

#### What did you do?

Clone the repo at https://github.com/kevinburke/go-mod-nondeterministic.

Running `GO111MODULE=on go mod tidy` modifies go.sum.
Running `GO111MODULE=on go mod vendor` changes go.sum back to the original value.

#### What did you expect to see?

I expect `go.sum` to reach a steady state at some point, where further command invocations make no changes to the file.

#### What did you see instead?

`go.sum` changes after running each of `go mod tidy` and `go mod vendor`. There's no "steady state" since the file state depends on the last command that was run.

#### Does this issue reproduce with the latest release (go1.11)?

Yes

#### System details

```
go version devel +93ad702251 Tue Sep 25 22:07:55 2018 +0000 darwin/amd64
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/kevin/Library/Caches/go-build"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOOS="darwin"
GOPATH="/Users/kevin"
GOPROXY=""
GORACE=""
GOROOT="/Users/kevin/go"
GOTMPDIR=""
GOTOOLDIR="/Users/kevin/go/pkg/tool/darwin_amd64"
GCCGO="gccgo"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD=""
GOROOT/bin/go version: go version devel +93ad702251 Tue Sep 25 22:07:55 2018 +0000 darwin/amd64
GOROOT/bin/go tool compile -V: compile version devel +93ad702251 Tue Sep 25 22:07:55 2018 +0000
uname -v: Darwin Kernel Version 17.7.0: Thu Jun 21 22:53:14 PDT 2018; root:xnu-4570.71.2~1/RELEASE_X86_64
ProductName:	Mac OS X
ProductVersion:	10.13.6
BuildVersion:	17G65
lldb --version: lldb-1000.11.37.1
  Swift-4.2
```

**Comment by myitcv** [comment]
The delta appears within `github.com/sourcegraph/zoekt` so this is almost certainly some function of the replace directive that exists in the `go.mod`:

```diff
diff --git a/go.sum b/go.sum
index 1d2e68e..034fb23 100644
--- a/go.sum
+++ b/go.sum
@@ -26,7 +26,6 @@ github.com/pelletier/go-buffruneio v0.2.0/go.mod h1:JkE26KsDizTr40EUHkXVtNPvgGtb
 github.com/pkg/errors v0.8.0/go.mod h1:bwawxfHBFNV+L2hUp1rHADufV3IMtnDRdf1r5NINEl0=
 github.com/pmezard/go-difflib v1.0.0/go.mod h1:iKH77koFhYxTK1pcRnkKkqfTogsbg7gZNVY4sRDYZ/4=
 github.com/sergi/go-diff v1.0.0/go.mod h1:0CfEIISq7TuYL3j771MWULgwwjU+GofnZX9QAmXWZgo=
-github.com/sourcegraph/zoekt v0.0.0-20180925141536-852b3842c11d h1:82xTRg5XBKlVsvLvgcgQzcG54AFTYhS2DdMVT5U5AWw=
 github.com/sourcegraph/zoekt v0.0.0-20180925141536-852b3842c11d/go.mod h1:f1Qnbu4glSRTb9A/H7qLv4AhudFf3eStkDbpyb21KYQ=
 github.com/src-d/gcfg v1.3.0/go.mod h1:p/UMsR43ujA89BJY9duynAwIpvqEujIH/jFlfL7jWoI=
 github.com/stretchr/testify v1.2.2/go.mod h1:a8OnRcib4nhh0OaRAV+Yts87kKdq0PP7pXfy6kDkUVs=
```

cc @rsc @bcmills

**Comment by kevinburke** [comment]
FWIW, I'm manually editing go.sum multiple times per day on a (soon to be, large, probably popular) open source project to work around this issue.

**Comment by odiferousmint** [comment]
I never run `go build` in the project's directory because it changes the `go.sum` and adds lots of duplicates of dependencies with a different checksum (IIRC), otherwise I would have to remove `go.sum` and do `go mod verify` every time I want to git push.

**Comment by dmitris** [comment]
I'm having an issue with `go.sum` being modified as a side-effect of running `go install ./...` (or `go mod why` - but **not** with `go mod tidy`):

**Comment by dmitris** [comment]
@thepudds asked on Slack to post the samples of go.sum - attaching 3 files: `go.mod.txt`, `go.sum.txt`, and `go.sum.modified.txt
 with modifications described above introduced by running `go install ./...` (or `go mod why -m <pkg>`).  Removed internal dependencies (not involved in the diffs) and "falsified" the mirror domain name with the fake `mirror.foobar.com` value.

[go.mod.txt](https://github.com/golang/go/files/2456260/go.mod.txt)
[go.sum.txt](https://github.com/golang/go/files/2456262/go.sum.txt)
[go.sum.modified.txt](https://github.com/golang/go/files/2456264/go.sum.modified.txt)

**Comment by myitcv** [comment]
`go mod tidy` appears to be at fault here (there may be other issues at play), but here's a reproduction:

```
$ export GOPATH=$(mktemp -d)
$ export PATH=$GOPATH/bin:$PATH
$ cd $(mktemp -d)
$ go mod init example.com/hello
go: creating new go.mod: module example.com/hello
$ cat <<EOD >main.go
package main
import _ "github.com/gopherjs/gopherjs/js"
import _ "rsc.io/quote"

func main() {}
EOD
$ go mod edit -require=github.com/gopherjs/gopherjs@latest -replace=github.com/gopherjs/gopherjs=github.com/myitcv/gopherjs@go1.11
$ go mod tidy
go: finding github.com/gopherjs/gopherjs latest
...
```

Look at the number of lines in `go.sum` for both `rsc.io/quote` (not replaced) and GopherJS (replaced):

```
$ cat go.sum | grep quote
rsc.io/quote v1.5.2 h1:w5fcysjrx7yqtD/aO+QwRjYZOKnaM9Uh2b40tElTs3Y=
rsc.io/quote v1.5.2/go.mod h1:LzX7hefJvL54yjefDEDHNONDjII0t9xZLPXsUe+TKr0=
$ cat go.sum | grep gopherjs
github.com/myitcv/gopherjs v0.0.0-20181007174709-d1c34fc540e3/go.mod h1:X83RgTaopDGeqahgXrQ120BVA5Ztn+oW4kxNueqOja4=
```

The single line for `github.com/myitcv/gopherjs` looks wrong.

"Fix" this using `go list all`:

```
$ go list all >/dev/null
go: finding github.com/myitcv/gopherjs v0.0.0-20181007174709-d1c34fc540e3
```

and check again:

```
$ cat go.sum | grep quote
rsc.io/quote v1.5.2 h1:w5fcysjrx7yqtD/aO+QwRjYZOKnaM9Uh2b40tElTs3Y=
rsc.io/quote v1.5.2/go.mod h1:LzX7hefJvL54yjefDEDHNONDjII0t9xZLPXsUe+TKr0=
$ cat go.sum | grep gopherjs
github.com/myitcv/gopherjs v0.0.0-20181007174709-d1c34fc540e3 h1:gHMnSFk1naYj5G8j9hhiYQ7kKWbsHjjGSPI8sM+Fqt4=
github.com/myitcv/gopherjs v0.0.0-20181007174709-d1c34fc540e3/go.mod h1:X83RgTaopDGeqahgXrQ120BVA5Ztn+oW4kxNueqOja4=
```

Run `go mod tidy` again to see if our "fix" persists:

```
$ go mod tidy
$ cat go.sum | grep quote
rsc.io/quote v1.5.2 h1:w5fcysjrx7yqtD/aO+QwRjYZOKnaM9Uh2b40tElTs3Y=
rsc.io/quote v1.5.2/go.mod h1:LzX7hefJvL54yjefDEDHNONDjII0t9xZLPXsUe+TKr0=
$ cat go.sum | grep gopherjs
github.com/myitcv/gopherjs v0.0.0-20181007174709-d1c34fc540e3/go.mod h1:X83RgTaopDGeqahgXrQ120BVA5Ztn+oW4kxNueqOja4=
```

It doesn't; hence we conclude `go mod tidy` is over aggressively pruning `go.sum`.

_The "fix" here could involve the use of any `go` command that effectively invokes a build step, hence install, build, list etc all work._

**Comment by dmitris** [comment]
@bcmills - do you (or anyone from the Go team) agree that `go mod tidy` needs to be rectified?  Is anyone working on this?  What are the chances this can be fixed in 1.12?

**Comment by dmitris** [comment]
maybe as a workaround it would be possible for `go mod tidy` to "internally" run an equivalent of `go list all >& /dev/null` to "effectively invoke a build step" and "untidy" its over-aggressive purning?
